File: Leetcode-Python/N-Queens2.py
"""
Follow up for N-Queens problem.

Now, instead outputting board configurations, return the total number of distinct solutions.
"""
import copy


class Solution:

    # @param {integer} n
    # @return {integer}
    def totalNQueens(self, n):
        self.__res = []
        self.__count = 0

        def __totalNQueensDFS(n, rows):
            if len(rows) == n:
                # NOTE: DEEP COPY!!!!!!!
                # here `rows` is a just a list of integers,
                # so can also copy like these:
                # temp = rows[:]
                # temp = copy.copy(rows)
                temp = copy.deepcopy(rows)
                self.__res.append(temp)
                self.__count += 1
            else:
                for i in xrange(n):
                    can_put_col_i = True  # can put at (len(rows), i)?
                    for idx, pos in enumerate(rows):
                        if not valid(idx, pos, len(rows), i):
                            can_put_col_i = False
                            break
                    if can_put_col_i:
                        rows.append(i)
                        __totalNQueensDFS(n, rows)
                        rows.pop()

        def valid(x1, y1, x2, y2):
            return y1 != y2 and abs(x2 - x1) != abs(y2 - y1)

        __totalNQueensDFS(n, [])
        return self.__count


if __name__ == '__main__':
    s = Solution()
    print s.totalNQueens(4)

File: Java-By-Tag/src/main/NQueens2.java
package src.main;

/**
 * Created by yhf on 10/10/15.
 */

import java.util.List;

/**
 * Follow up for N-Queens problem.
 * <p>
 * Now, instead outputting board configurations, return the total number of distinct solutions.
 */

public class NQueens2 {

    private int count = 0;

    public int totalNQueens(int n) {
        if (n <= 0) {
            return 0;
        }
        totalNQueens(n, new int[n], 0);
        return count;
    }

    private void totalNQueens(int n, int[] rows, int row) {
        if (row == n) {
            count++;
        }

        for (int colIdx = 0; colIdx < rows.length; colIdx++) {
            boolean existsInvalid = false;
            for (int rowIdx = 0; rowIdx < row; rowIdx++) {
                if (!isValid(rowIdx, rows[rowIdx], row, colIdx)) {
                    existsInvalid = true;
                    break;
                }
            }
            if (!existsInvalid) {
                rows[row] = colIdx;
                totalNQueens(n, rows, row + 1);
                rows[row] = 0;
            }
        }
    }

    private boolean isValid(int x1, int y1, int x2, int y2) {
        return x1 != x2 && y1 != y2 && Math.abs(x1 - x2) != Math.abs(y1 - y2);
    }
}
